operator/k8s_identity.go

```diff
@@ -0,0 +1,116 @@
+// Copyright 2019 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"time"
+
+	"github.com/cilium/cilium/pkg/controller"
+	"github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
+	"github.com/cilium/cilium/pkg/k8s/informer"
+	"github.com/cilium/cilium/pkg/k8s/types"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+)
+
+var identityStore cache.Store
+
+func deleteIdentity(identity *types.Identity) error {
+	err := ciliumK8sClient.CiliumV2().CiliumIdentities().Delete(
+		identity.Name,
+		&metav1.DeleteOptions{
+			Preconditions: &metav1.Preconditions{
+				UID:             &identity.UID,
+				ResourceVersion: &identity.ResourceVersion,
+			},
+		})
+	if err != nil {
+		log.WithError(err).Error("Unable to delete identity")
+	}
+
+	return err
+}
+
+func identityGCIteration() {
+	if identityStore == nil {
+		return
+	}
+
+nextIdentity:
+	for _, identityObject := range identityStore.List() {
```

Comment:
Is it possible that after we do the `List` of an Identity has `len(identity.Status.Nodes) == 0`, that a K8s client in the cluster has then added the identity again? In other words, are we safe guarded against resource version changes for an identity after it has been listed here, but then updated shortly after `identityStore.List()` has been performed?

Reply:
It is guarded. `deleteIdentity` (called below) includes the `ResourceRevision` to the Delete call. That should return an error if the version doesn't match. It would log this and no-op. In the case of GC, this is benign and it will be retried in the next iteration.
Normally, identities would be deleted as soon as the final node removes itself. In those cases, the delete failing would leave the CRD instance for GC to clean-up. Both cases could race with new users of an identity. Generally, they would recreate the "master" key, in this case just the CRD, if an in-use identity is deleted. This is part of `pkg/allocator.Allocator`.

In any case, I added some more comments to try and make this more clear.`